### DISCUSSION TITLE
Sample App Proposal about EuPI-Camera demo app

### DISCUSSION THREAD

**Discussion by SeonJK** [euphony-io/euphony#209] [Ideas]
Hi, there!
I and @yoouyoou want to make EuPI-Camera demo app.

<br/>

# Apps
we will make two apps.

## >Camera app
- use CameraX library in Android Jetpack
- when listen EuPI, do some features
## >Remote control app
- have three buttons to send EuPI (*in progress*)
  - send EuPI for camera capture
  - send EuPI for camera swtich
  - send EuPI for start/stop to record video
- use Compose

**Comment by designe** [comment]
Hi! @SeonJK @yoouyoou 
I'm looking forward to seeing the interesting `EuPI` sample app :)
I think you need to develop 2 apps in single android project.

Let me share some links how to make multi apps in single project.
- https://www.codexpedia.com/android/android-multiple-apps-and-modules-in-one-project-demo/
- https://medium.com/swlh/managing-multiple-apps-within-a-single-android-studio-project-659eb33f5f0e

Anyway, Thanks for sharing your proposal in decussions 👍

**Reply by SeonJK** [reply]
Thank you for links!
AndI have a question for you!

what's the difference between multi-apps and two project in a repository?

**Reply by designe** [reply]
I'd like to explain with an example.
Basically, when you create a project, it consists of build.gradle at the root level of the project and build.gradle for module.
```
{$project_root_level}/build.gradle
{$project_root_level}/settings.gradle
{$project_root_level}/{$module_level}/build.gradle
```
How Android Studio recognizes a project is looking at `build.gradle` and `settings.gradle` in $project_root_level. 
Android Studio can run one project in one IDE.

If you create two projects here
```
project_1/build.gradle
project_1/settings.gradle
project_1/{$module_level}/build.gradle
project_2/build.gradle
project_2/settings.gradle
project_2/{$module_level}/build.gradle
```
You need to run and manage 2 IDEs.

On the other hand, if you use multi module, you can manage multiple apps in one IDE because one project is used.

```
{$project_root_level}/build.gradle
{$project_root_level}/settings.gradle
{$project_root_level}/app/build.gradle
{$project_root_level}/app2/build.gradle
{$project_root_level}/app3/build.gradle
///...
```

To give an example of parallel processing, I think it is the difference between whether it is implemented as a multi-process or multi-threaded implementation. The former(multi-project) is multi-process and the latter(multi-modules) is multi-threaded.

Below is good application of multi-modules.
https://github.com/google/ExoPlayer/blob/release-v2/settings.gradle

**Reply by SeonJK** [reply]
multi-modules seem to share `build.gradle` and `settings.gradle` in project_root_level.
I have to study more about that....lol

Thanks for your answer!!

**Reply by yeonns** [reply]
@SeonJK Don't have to feel difficult :)
You can just make new project with android studio, and make new module with `File - New - New module - Phone & Tablet`.
Then, you will have multi-module like this :smile: 
- app (initial project)
- test-app (make new module with `File - New - New module`)
![스크린샷, 2022-08-31 22-02-45](https://user-images.githubusercontent.com/10149398/187684705-c144fcd8-3fff-4096-9890-fa7edfd0e7af.png)

After making module, you can also rename module :)
- right click `app` directory -> Refactor -> Rename -> Rename module (ex. eupi-camera)

**Reply by yeonns** [reply]
@designe 
If we have multi modules, what is the best naming of repository and each modules? :eyes: 
e.g.)
- repo name: `eupi-sample`
  - module1: `eupi-camera`
  - module2: `eupi-remote-control`

**Reply by designe** [reply]
How about this?

How about naming it like the one below following the `wave-in` series?
- `wave-in-listener` : https://github.com/euphony-io/wave-in-listener
- `wave-in-speaker` : https://github.com/euphony-io/wave-in-speaker

- repo name (project name): `wave-cam`
  - module1: `eupi-camera`
  - module2: `eupi-shutter`

**Reply by yeonns** [reply]
@designe Oh, good idea :+1: 
@SeonJK @yoouyoou how do you think about repo and module name that @designe suggested? :smile:

**Reply by yoouyoou** [reply]
The name of the wave series repository is good!

We thought like this.
module 1: eupi-camera-transmit
module 2: eupi-camera-receive

How about this?
@SeonJK  I'll have to discuss more.

**Reply by designe** [reply]
one is the camera app, and the other app is shutter-release role.
@SeonJK @yoouyoou you are the owner of those one :) hope to make funny name 😃

**Comment by SeonJK** [comment]
@yeonns @designe 
`wave-cam` is good for repo name!! we will change this 👍

Thanks for suggestions :)
